mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp

```diff
@@ -1145,21 +1148,399 @@ ReduceRewritePattern::matchAndRewrite(rock::ReduceOp reduceOp,
+
+  // Currently, we use blockwise_reductions when fusing with reductions
+  // if and only if all sub-tile view creations are successful.
+  if (succeeded(blockSubTileViews) && succeeded(threadSubTileViews) &&
+      succeeded(blockSubTileTidSliceViews) && succeeded(gridOnlyDims) &&
+      succeeded(lowerSubDims)) {
+    ArrayRef<int64_t> blockLowerShape =
+        getLowerShape(blockSubTileViews.value());
+    ArrayRef<int64_t> blockSubTileTidSliceShape =
+        getLowerShape(blockSubTileTidSliceViews.value());
+    int64_t blockSubTileTidSliceRank = blockSubTileTidSliceShape.size();
+    // The block sub-tile view might not have the slower changing
+    // dimensions in it. Thus, we always keep track of the reduction
+    // dimensions from its distance to fastest changing dimensions.
+    blockReductionAxis =
+        blockSubTileTidSliceRank - 1 - blockReductionAxisFromLeft;
+    int64_t partialReductionsPerThread =
+        blockSubTileTidSliceShape[blockReductionAxis];
+    int64_t ldsWorkspaceSize = 1;
+    for (auto [idx, size] : llvm::enumerate(blockLowerShape)) {
+      if (idx == (size_t)blockReductionAxis) {
+        ldsWorkspaceSize *= partialReductionsPerThread;
+      } else {
+        ldsWorkspaceSize *= size;
+      }
     }
+    TypedValue<MemRefType> src = threadwiseWriteOp.getSource();
+    auto broadcastReducedSrc = rewriter.create<GpuAllocOp>(loc, src.getType());
+    Value ldsWorkspace = rock::gpuAlloc(rewriter, loc, ldsWorkspaceSize,
+                                        src.getType().getElementType(),
+                                        gpu::AddressSpace::Workgroup);
+
+    rewriter.create<BlockwiseBroadcastReduceOp>(
+        loc, src, ldsWorkspace, broadcastReducedSrc,
+        /*extraOut=*/nullptr, rewriter.getIndexAttr(blockReductionAxis),
+        reduceOp.getReduceMethodAttr(), blockSubTileViews.value(),
+        blockSubTileTidSliceViews.value(), threadSubTileViews.value(),
+        /*extraViews=*/nullptr,
+        getBlockSize(reduceOp->getParentOfType<func::FuncOp>()).value());
+
+    ViewLikeOpInterface viewOp =
+        ldsWorkspace.getDefiningOp<ViewLikeOpInterface>();
+    rewriter.create<GpuDeallocOp>(loc, viewOp.getViewSource());
+    // Create partial reduction views
+    ArrayAttr paddedReducedTrStack;
+    {
+      SmallVector<Attribute> transformAttrs;
+      ArrayRef<int64_t> blockTileShape =
+          getLowerShape(blockSubTileViews.value());
+      SmallVector<SmallString<8>> names;
+      SmallVector<StringRef> nameRefs;
+      for (size_t i = 0; i < blockTileShape.size(); i++) {
+        SmallString<8> dimName(Twine("dim" + Twine(i)).str());
+        names.push_back(dimName);
+        nameRefs.push_back(names.back());
+      }
+      TopDownTMBuilder toReducedView(rewriter, nameRefs, blockTileShape);
+      for (unsigned i = 0; i < blockTileShape.size(); i++) {
+        if (blockReductionAxis == i) {
+          toReducedView.pad({nameRefs[i]}, {0, blockTileShape[i] - 1});
```

Comment:
... hold on, is this using `Pad{}` to take a length-1 dimension to a length-`blockTileShape[i]` one?

This feels like it could be `Broadcast{}`?